sound/soc/intel/boards/sof_es8336.c

```diff
@@ -722,30 +703,64 @@ static int sof_es8336_probe(struct platform_device *pdev)
 		}
 	}
 
-	/* get speaker enable GPIO */
-	if (quirk & SOF_ES8336_HEADPHONE_GPIO) {
-		if (quirk & SOF_ES8336_SPEAKERS_EN_GPIO1_QUIRK)
-			gpio_mapping = acpi_enable_both_gpios;
-		else
-			gpio_mapping = acpi_enable_both_gpios_rev_order;
-	} else if (quirk & SOF_ES8336_SPEAKERS_EN_GPIO1_QUIRK) {
-		gpio_mapping = acpi_speakers_enable_gpio1;
+	if (quirk & SOF_ES8336_SPEAKERS_EN_GPIO1_QUIRK) {
+		priv->enable_spk_gpio.crs_entry_index = 1;
+		priv->enable_hp_gpio.crs_entry_index = 0;
+	} else {
+		priv->enable_spk_gpio.crs_entry_index = 0;
+		priv->enable_hp_gpio.crs_entry_index = 1;
+	}
```

Comment:
I am lost on this one @mchehab, do we actually need a quirk at all for the speaker GPIOs? Why not assume that the speakers always has a GPIO and simplify - which was the recommendation from @yangxiaohua2009

Reply:
The patch is assuming that speakers and headphones will always have GPIOs. The point is that, on Matebook D15, the speaker GPIO is 1, while the headphone GPIO is 0. Other devices could have this reversed - and this was the default before we added proper support for it.

While I would love to, I can't see a way to avoid this quirk - at least for now. Perhaps there are some information somewhere else at the ACPI tables telling what's connected to each GPIO port, but I was unable to identify where.

Reply:
Can you tell us the value of GPIO0 when speaker on? If the value of GPIO0 is low when speaker on and you can hear sound from speaker then a quirk is need.

Reply:
The patch is assuming that speakers and headphones will always have GPIOs. The point is that, on Matebook D15, the speaker GPIO is 1, while the headphone GPIO is 0. Other devices could have this reversed - and this was the default before we added proper support for it.

While I would love to, I can't see a way to avoid this quirk - at least for now. Perhaps there are some information somewhere else at the ACPI tables telling what's connected to each GPIO port, but I was unable to identify where.

> Can you tell us the value of GPIO0 when speaker on? If the value of GPIO0 is low when speaker on and you can hear sound from speaker then a quirk is need.

Both GPIOs need to be changed the same way, as one is low triggered, while the other is high triggered. So, turning speaker on /headphone off means setting both to true, and turning speaker off/headphone on means setting both to false.

Reply:
I have a Huawei AMD laptop with es8336, and I wrote a machine driver for it which I plan to upstream. I cherry picked these commits to add support for this DSM parsing. The GPIO stuff is not working for me. The DSM code reads that GPIO0 is active high and GPIO1 is active low:

`[   47.483034] es8316 i2c-ESSX8336:00: speaker gpio 0 active high, headphone gpio 1 active low`

Now when running on speaker everything is ok, but whenever I plug in the headset, no sound can be heard in the headphones or the speaker. If I comment out the second the call to `gpiod_set_value_cansleep(priv->gpio_headphone, !priv->speaker_en);`

The speaker/headset switching and back works ok.

To me it seems that it is not always safe to use both gpios (at least on AMD platforms).

Maybe there is a way to detect in DSM if this system uses one or both GPIOs? I saw this entry which sounds interesting: `CODEC_GPIO0_FUNC_ARG=0x1`

Reply:
The headphone gpio dsm returns 0xff (default value), which should be active high instead of active low. This is a known issue which needs bo be fixed.

Reply:
> The headphone gpio dsm returns 0xff (default value), which should be active high instead of active low. This is a known issue which needs bo be fixed.

Such logic works fine for Matebook D15, so it seems that we'll end needing a quirks anyway for such notebook. Btw, I quickly tested on a Matebook D14. Despite ACPI looking identical, the logic there seems to be inverted, but I need to do more tests.

In summary, it seems that we'll keep needing a quirk to override DSM. Unfortunately, I lost access to both notebooks, and I don't know when I'll be able to access them again. I'm hoping to get my hands again on Matebook D14 in Feb, and work on another version of this patch series.

Reply:
@plbossart we should have a fuction like this:
```
switch (value) {
	case GPIO_CTL_IO_LEVEL_HIGH:
       	case 0xff:
		return 0;
	case GPIO_CTL_IO_LEVEL_LOW:
		return 1;
	default:
		dev_err(dev, "%s: invalid I/O ctl level argument (%d)\n",
			type, value);
		return -EINVAL;
	}
}
```

instead of 

```
switch (value) {
	case GPIO_CTL_IO_LEVEL_HIGH:
		return 0;
	case GPIO_CTL_IO_LEVEL_LOW:
	case 0xff:
		return 1;
	default:
		dev_err(dev, "%s: invalid I/O ctl level argument (%d)\n",
			type, value);
		return -EINVAL;
	}
}
```

Actually BIOS can be changed by Windows Updates but this may not work for Linux.